massdriver/resource_package_alarm_test.go

```diff
@@ -28,6 +34,27 @@ func testAccCheckMassdriverPackageAlarmConfigBasic() string {
 	resource "massdriver_package_alarm" "new" {
 		cloud_resource_id = "arn:::something"
 		display_name = "CPU alarm"
+		metric {
+			name = "Metric Name"
+			namespace = "Metric/Namespace"
+			statistic = "SUM"
+			dimensions = {
```

Comment:
non-blocking!
Could we make this strongly typed?
```
[
  {
      name = "dimension1"
      value = "abc"
   }
   {
     name = "dimension2"
     value = "xyz"
   }
]
```

That'd be closer to what the payloads we generally process for alarms look like:
https://github.com/massdriver-cloud/massdriver/blob/main/test/support/cloud_alarm_case.ex#L26
https://github.com/massdriver-cloud/massdriver/blob/main/test/support/cloud_alarm_case.ex#L210

Reply:
The dimensions field is already strongly typed by this: https://github.com/massdriver-cloud/terraform-provider-massdriver/pull/22/files#diff-ac54506c949ba8a02ebd2f01d0b06507bd46f9935950760e2597589cc756b2f3R15
Go enforces it to be a `map[string]string` in how it will be serialized to JSON.

The terraform itself is also effectively a `map[string]string`: https://github.com/massdriver-cloud/terraform-provider-massdriver/pull/22/files#diff-ac54506c949ba8a02ebd2f01d0b06507bd46f9935950760e2597589cc756b2f3R67-R74
Which matches the way that the AWS cloudwatch alarm resource manages the field: https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/cloudwatch/metric_alarm.go#L94-L98
Which allows us to do stuff like this:

```
resource massdriver_package_alarm "foo" {
  ...
  metric {
    ...
    dimensions = aws_cloudwatch_alarm.foo.metric.dimensions
  }
}
```

If transformations to the data need to happen I think it's best to keep it isolated to the metrics processing system and not extend its requirements or peculiarities out to connected systems. I'll defer to @xpositivityx .

Reply:
my gut feeling and I am open to hear other opinions is that storing this as a list of maps is an idiosyncrasy of ecto and our access patterns. When thinking in terms of the message being sent as json that may one day be consumed by many services, implementing a list of maps instead of a map feels like that implementation detail is being distributed to all services. If there is a potential however, for there to be multiple overlapping dimension names, the map form of this would break but that would also be a real problem for terraform state.

Reply:
That makes sense to me.

Thanks for the feedback, all!